File: leetcode/binary-tree-preorder-traversal/binary-tree-preorder-traversal-iter.rb
# LeetCode #144 - Binary Tree Preorder Traversal
# https://leetcode.com/problems/binary-tree-preorder-traversal/
# Simple iterative solution with a node stack.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def preorder_traversal(root)
  out = []
  stack = []
  stack << root if root

  until stack.empty?
    node = stack.pop
    out << node.val
    stack << node.right if node.right
    stack << node.left if node.left
  end

  out
end

File: leetcode/binary-tree-preorder-traversal/binary-tree-preorder-traversal-iter.java
// LeetCode #114 - Binary Tree Preorder Traversal
// https://leetcode.com/problems/binary-tree-preorder-traversal/
// Iterative implementation.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        preorder(root, values::add);
        return values;
    }

    private static void preorder(TreeNode root, IntConsumer action) {
        if (root == null) return;

        var stack = Collections.asLifoQueue(new ArrayDeque<TreeNode>());
        stack.add(root);

        while (!stack.isEmpty()) {
            var node = stack.remove();
            action.accept(node.val);
            if (node.right != null) stack.add(node.right);
            if (node.left != null) stack.add(node.left);
        }
    }
}
